### ISSUE TITLE
aw-watcher-afk/aw-watcher-window can not connect to the aw-server

### ISSUE THREAD

**Issue by v3470** [ActivityWatch/activitywatch#442]
- [x] I am on the [latest](https://github.com/ActivityWatch/activitywatch/releases/latest) ActivityWatch version.
 - [x] I have searched the issues of this repo and believe that this is not a duplicate.

- **OS version and name**: Windows 10 Home Edition, version 10.0.18363.900
- **ActivityWatch version**: 0.9.2

## Describe the bug

aw-watcher-afk and aw-watcher-window can not connect to the aw-server. The aw-server is running and the browser extension can connect to it. Also, there can not query the activity summary of period after this problem happened, but the Timeline view can still show current event log.

## To Reproduce

Maybe shutdown the computer casually can cause this problem.

## Expected behavior

aw-watcher-* work regularly and the Activity view of the Dashboard can query current summary.

## Documentation

The Activity view:
![图片](https://user-images.githubusercontent.com/58453657/84864774-c1773c80-b0a9-11ea-8d7b-9c567e176d9c.png)
The Timeline view:
![图片](https://user-images.githubusercontent.com/58453657/84864809-d0f68580-b0a9-11ea-8214-99e0f43eab23.png)

Below is the recent aw-server's logs:
```
2020-06-17 12:58:22 [INFO ]: Using storage method: peewee  (aw_server.main:26)
2020-06-17 12:58:22 [INFO ]: Starting up...  (aw_server.main:31)
2020-06-17 12:58:22 [INFO ]: Using database file: C:\Users\user\AppData\Local\activitywatch\activitywatch\aw-server\peewee-sqlite.v2.db  (aw_datastore.storages.peewee:90)
2020-06-17 12:58:22 [INFO ]:  * Running on http://localhost:5600/ (Press CTRL+C to quit)  (werkzeug:122)
2020-06-17 12:58:22 [INFO ]: Received heartbeat after pulse window, inserting as new event. (bucket: aw-watcher-web-firefox)  (aw_server.api:228)
2020-06-17 12:58:25 [INFO ]: Received heartbeat after pulse window, inserting as new event. (bucket: aw-watcher-web-firefox)  (aw_server.api:228)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration but could be safely merged (-50.146s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration but could be safely merged (-5259.732s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration and could NOT be safely merged (-5259.732s). This warning will only show once per batch.  (aw_transform.flood:50)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration but could be safely merged (-561.55s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration but could be safely merged (-539.827s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration but could be safely merged (-715.669s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration but could be safely merged (-1827.031s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:23 [WARNING]: Gap was of negative duration but could be safely merged (-1715.477s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:24 [WARNING]: Gap was of negative duration but could be safely merged (-0.01s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:24 [WARNING]: Gap was of negative duration but could be safely merged (-1639.936s). This message will only show once per batch.  (aw_transform.flood:46)
2020-06-17 14:25:24 [WARNING]: Gap was of negative duration and could NOT be safely merged (-1.397s). This warning will only show once per batch.  (aw_transform.flood:50)
```
and aw-qt's logs:
```
2020-06-17 12:58:21 [INFO ]: Starting module aw-server  (aw_qt.manager:49)
2020-06-17 12:58:21 [INFO ]: Starting module aw-watcher-window  (aw_qt.manager:49)
2020-06-17 12:58:21 [INFO ]: Starting module aw-watcher-afk  (aw_qt.manager:49)
2020-06-17 12:58:21 [INFO ]: Creating trayicon...  (aw_qt.trayicon:140)
```
and aw-watcher-afk's logs:
```
2020-06-17 12:58:21 [INFO ]: aw-watcher-afk started  (aw_watcher_afk.afk:53)
2020-06-17 12:58:22 [WARNING]: Not connected to server, 13906 requests in queue  (aw_client.client:380)
2020-06-17 12:58:32 [WARNING]: Not connected to server, 13906 requests in queue  (aw_client.client:380)
2020-06-17 12:58:42 [WARNING]: Not connected to server, 13907 requests in queue  (aw_client.client:380)
2020-06-17 12:58:53 [WARNING]: Not connected to server, 13908 requests in queue  (aw_client.client:380)
2020-06-17 12:59:03 [WARNING]: Not connected to server, 13909 requests in queue  (aw_client.client:380)
2020-06-17 12:59:13 [WARNING]: Not connected to server, 13909 requests in queue  (aw_client.client:380)
2020-06-17 12:59:23 [WARNING]: Not connected to server, 13910 requests in queue  (aw_client.client:380)
2020-06-17 12:59:33 [WARNING]: Not connected to server, 13911 requests in queue  (aw_client.client:380)
2020-06-17 12:59:43 [WARNING]: Not connected to server, 13912 requests in queue  (aw_client.client:380)
```
and aw-watcher-window's logs:
```
2020-06-17 12:58:21 [INFO ]: aw-watcher-window started  (aw_watcher_window.main:40)
2020-06-17 12:58:22 [WARNING]: Not connected to server, 11839 requests in queue  (aw_client.client:380)
2020-06-17 12:58:32 [WARNING]: Not connected to server, 11841 requests in queue  (aw_client.client:380)
2020-06-17 12:58:42 [WARNING]: Not connected to server, 11845 requests in queue  (aw_client.client:380)
2020-06-17 12:58:53 [WARNING]: Not connected to server, 11848 requests in queue  (aw_client.client:380)
2020-06-17 12:59:03 [WARNING]: Not connected to server, 11850 requests in queue  (aw_client.client:380)
2020-06-17 12:59:13 [WARNING]: Not connected to server, 11853 requests in queue  (aw_client.client:380)
2020-06-17 12:59:23 [WARNING]: Not connected to server, 11856 requests in queue  (aw_client.client:380)
2020-06-17 12:59:33 [WARNING]: Not connected to server, 11860 requests in queue  (aw_client.client:380)
2020-06-17 12:59:43 [WARNING]: Not connected to server, 11862 requests in queue  (aw_client.client:380)
2020-06-17 12:59:53 [WARNING]: Not connected to server, 11865 requests in queue  (aw_client.client:380)
2020-06-17 12:59:58 [ERROR]: Exception thrown while trying to get active window: (-2147217385, 'OLE error 0x80041017', None, None)  (aw_watcher_window.main:66)
2020-06-17 12:59:59 [ERROR]: Exception thrown while trying to get active window: (-2147217385, 'OLE error 0x80041017', None, None)  (aw_watcher_window.main:66)
2020-06-17 13:00:00 [ERROR]: Exception thrown while trying to get active window: (-2147217385, 'OLE error 0x80041017', None, None)  (aw_watcher_window.main:66)
2020-06-17 13:00:01 [ERROR]: Exception thrown while trying to get active window: (-2147217385, 'OLE error 0x80041017', None, None)  (aw_watcher_window.main:66)
2020-06-17 13:00:03 [WARNING]: Not connected to server, 11868 requests in queue  (aw_client.client:380)
2020-06-17 13:00:13 [WARNING]: Not connected to server, 11869 requests in queue  (aw_client.client:380)
2020-06-17 13:00:23 [WARNING]: Not connected to server, 11869 requests in queue  (aw_client.client:380)
```

## Additional context

I checked the sqlite database, the event is still in recording and I think it is still heathy.

**Comment by johan-bjareholt** [comment]
As you can see in the logs the window and afk watchers tries to attempt to connect to the server but fails to do so.
It's interesting that the browser watcher succeeds but the afk and window watcher does not, might it be something with your network settings which the browser watcher ignores while the others respect? (maybe a proxy or something?)

**Comment by v3470** [comment]
@johan-bjareholt I tried to remove the `http_proxy` env variable, and then the watchers connected to the server well. However, there was still no data show in the Activity, also the aw-watcher-afk and the aw-watcher-window did not log event on the Timeline(while the browser extension did log).

Tried to turn-off the system firewall doesn't make any effect, because the `http_proxy` in settings(I set it back after previous test), the watchers still log can not connect to the server.

I think it is a server side problem and have no a idea how to fix it because the can not find any detail of Python side traceback.

**Comment by johan-bjareholt** [comment]
> I tried to remove the http_proxy env variable, and then the watchers connected to the server well. However, there was still no data show in the Activity, also the aw-watcher-afk and the aw-watcher-window did not log event on the Timeline(while the browser extension did log).

The watchers queue up events if they fail to get sent, it's not unlikely that it is slowly importing days (weeks?) old activity and is therefore not yet visible in the timeline. Take a look if the amount of events in the buckets are increasing or look at the activity view and look at for example "month" and see if active time increases after a while.

> I think it is a server side problem and have no a idea how to fix it because the can not find any detail of Python side traceback.

There's no guarantee that there even will be a traceback as those only occur if there some kind of panic, raised exception or crash.

**Comment by v3470** [comment]
> Take a look if the amount of events in the buckets are increasing or look at the activity view and look at for example "month" and see if active time increases after a while.

The aw-watcher-afk_local and aw-watcher-window_local buckets were not increasing, as the activity view of month/week/day were not update since June 15. Is there anything else I can do for find the reason of this problem?

**Comment by johan-bjareholt** [comment]
> The aw-watcher-afk_local and aw-watcher-window_local buckets were not increasing, as the activity view of month/week/day were not update since June 15. Is there anything else I can do for find the reason of this problem?

And this was with http_server env variable unset?

If yes, how are you sure that that the clients/watchers properly connected?

**Comment by v3470** [comment]
> And this was with http_server env variable unset?

In June 19th, I stoped the ActivityWatch, removed the http_proxy env variable and then restart the ActivityWatch and wait for about 6 mintues to see whether the new events come. At that time, the log of watchers show that connection has established at start. Because this was just a temporary testing, I set back the http_proxy after 6 minutes. During the 6 minutes, no new event update.

**Comment by johan-bjareholt** [comment]
@v3470 Did you look at the event count or just at the event list and saw that it was empty?
Since you had a queue at 11000 and 13000 heartbeats from your aw-watcher-window and aw-watcher-afk watchers according to the logs and they send the oldest hearbeat first it would only be visible on the eventcount and not on the event list as the list only show the last 1h. It would likely take at least a few hours for that queue to get emptied and your activity to catch up, probably more.

If you did look at the event count and the heartbeats got properly sent to aw-server, then you have to look at the aw-server logs from that time in hopes that there's some error printed.

**Comment by v3470** [comment]
@johan-bjareholt your guess is true. After the ActivityWatch running for few hours without http_proxy env variable, the events come up and now it works fine. For now there is another problem is that how can I regularly use ActivityWatch with a http_proxy env variable? This env variable config is important for daily use personally.

**Comment by ErikBjare** [comment]
@v3470 There should be some way to unset it specifically for ActivityWatch on Windows. 

I searched and found that this command should do it `set HTTP_PROXY=` (although I didn't test) so you'll have to write a script or something that unsets the variable before starting ActivityWatch. More details here: https://stackoverflow.com/a/13222794/965332

**Comment by v3470** [comment]
@ErikBjare OK thanks, it is a way to do that. But why do you think this case isn't a bug?

**Comment by ErikBjare** [comment]
I don't think it's a bug because it's actually doing the right thing by respecting the http_proxy environment variable. It's technically not an issue with ActivityWatch, but with the environment it runs in. 

We could detect that the environment variable is set and emit a warning. Feel free to open a separate issue (or submit a PR) for that, if you think it'd be helpful to others.

**Comment by ErikBjare** [comment]
You might also be able to fix it by setting the variable `NO_PROXY` like this: `SET NO_PROXY=localhost,127.0.0.1`

You could also set it globally, if you want.

**Comment by johan-bjareholt** [comment]
@ErikBjare while true, it could be same to make the default to not respect the http_proxy command as we don't expect people to want to expose their aw-watchers outside of localhost. We could for example have a config for it in aw-client to toggle it on/off.
